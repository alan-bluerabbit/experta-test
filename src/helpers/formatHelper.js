const formatHelper = (provider) => {
    const errors = {
        name: provider.name === '',
        mail: !validateEmail(provider.mail),
        cuit: !validateNumbers(provider.cuit),
        address: provider.name === '',
        phone: !validateNumbers(provider.phone),
    }
    return errors
}

const validateEmail = (email) => {
    const re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(String(email).toLowerCase());
}

const validateNumbers = (number) => {
    const re = /^[0-9\.\-\/]+$/;
    return re.test(String(number).toLowerCase());
}

export { formatHelper }