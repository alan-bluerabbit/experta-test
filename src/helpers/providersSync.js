import firebase from 'firebase'
import firebaseConfig from '../config/firebase'

// Initialize Firebase
firebase.initializeApp(firebaseConfig)

export default {
    getAllProviders: async () => {
        return firebase
            .firestore()
            .collection('providers')
            .get().then(query => {
                let providers = []
                query.forEach(doc => {
                    providers.push({
                        id: doc.id,
                        ...doc.data()
                    })
                });
                return providers
            })
    },
    createProvider: async (provider) => {
        const providerWithTimeStamp = {
            ...provider,
            created: firebase.firestore.Timestamp.now().seconds,
        }
        return firebase
            .firestore()
            .collection('providers')
            .add(providerWithTimeStamp)
    },
    updateProvider: async (provider) => {
        return firebase
            .firestore()
            .collection('providers')
            .doc(`${provider.id}`)
            .update({
                name: provider.name,
                mail: provider.mail,
                address: provider.address,
                cuit: provider.cuit,
                phone: provider.phone,
                resinc: provider.resinc
            });
    },
    deleteProvider: (providerId) => {
        return firebase
          .firestore()
          .collection('providers')
          .doc(`${providerId}`)
          .delete()
      },
}