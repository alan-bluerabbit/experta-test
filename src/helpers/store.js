import React, { createContext, useContext, useReducer } from 'react';

const StoreContext = createContext();
const initialState = {
  providers: []
};

const loadProviders = (state, action) => {
  let newProviders = action.providers
  return {
    ...state,
    providers: newProviders
  }
}

const reducer = (state, action) => {
  switch(action.type) {
    case "load":
      return loadProviders(state, action)
    default:
      throw new Error(`Unhandled action type: ${action.type}`);
  }
}

export const StoreProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <StoreContext.Provider value={{state, dispatch}}>
      {children}
    </StoreContext.Provider>
  )
}

export const useStore = () => useContext(StoreContext);