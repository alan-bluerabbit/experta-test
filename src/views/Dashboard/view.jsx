import React from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import Box from '@material-ui/core/Box';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import IconButton from '@material-ui/core/IconButton';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import AddIcon from '@material-ui/icons/Add';

import Providers from '../../Components/Providers';
import Copyright from '../../Components/Copyright'
import ProviderDialog from '../../Components/ProviderDialog';
import CreateDialog from '../../Components/CreateDialog';
import DeleteDialog from '../../Components/DeleteDialog';

import providersSync from '../../helpers/providersSync';
import { useStore } from '../../helpers/store'

import useStyles from './styles.js';

function Dashboard() {
	const { state, dispatch } = useStore()
	const [providerOpen, setProviderOpen] = React.useState(false);
	const [newProviderOpen, setNewProviderOpen] = React.useState(false);
	const [deleteProviderOpen, setDeleteProviderOpen] = React.useState(false);
	const [selectedProvider, setSelectedProvider] = React.useState({})
	React.useEffect(() => {
		getAllProviders()
	}, [])

	const classes = useStyles();

	const getAllProviders = async () => {
		const providers = await providersSync.getAllProviders()
		dispatch({type: 'load', providers})
	}
	
	const handleProviderOpen = (provider) => {
		setSelectedProvider(provider)
		setProviderOpen(true);
	};

	const handleProviderClose = () => {
		setSelectedProvider({})
		setProviderOpen(false);
	};

	const handleNewProviderOpen = () => {
		setNewProviderOpen(true);
	};

	const handleNewProviderClose = () => {
		setNewProviderOpen(false);
	};

	const handleDeleteProviderOpen = (provider) => {
		setSelectedProvider(provider)
		setDeleteProviderOpen(true);
	};

	const handleDeleteProviderClose = () => {
		setSelectedProvider({})
		setDeleteProviderOpen(false);
	};

	const handleAddProvider = async (newProvider) => {
		await providersSync.createProvider(newProvider)
		getAllProviders()
		handleNewProviderClose()
	}

	const handleEditProvider = async (updatedProvider) => {
		await providersSync.updateProvider(updatedProvider)
		getAllProviders()
		handleProviderClose()
	}

	const handleDeleteProvider = async () => {
		await providersSync.deleteProvider(selectedProvider.id)
		getAllProviders()
		handleDeleteProviderClose();
	}

	return (
		<div className={classes.root}>
			<CssBaseline />
			<AppBar position="absolute" className={classes.appBar}>
				<Toolbar className={classes.toolbar}>
						<Typography component="h1" variant="h6" color="inherit" noWrap className={classes.title}>
							Experta Seguros
						</Typography>
						<IconButton color="inherit" onClick={handleNewProviderOpen}>
							<AddIcon />
						</IconButton>
				</Toolbar>
			</AppBar>
			<main className={classes.content}>
			<div className={classes.appBarSpacer} />
			<Container maxWidth="lg" className={classes.container}>
					<Grid container spacing={3}>
						<Grid item xs={12}>
								<Paper className={classes.paper}>
								<Providers
									providers={state.providers}
									openProviderDialog={handleProviderOpen}
									deleteProvider={handleDeleteProviderOpen}
									reload={getAllProviders}
								/>
								</Paper>
						</Grid>
					</Grid>
					<Box pt={4}>
						<Copyright />
					</Box>
			</Container>
			</main>
			<ProviderDialog provider={selectedProvider} open={providerOpen} handleClose={handleProviderClose} editProvider={handleEditProvider}/>
			<CreateDialog open={newProviderOpen} handleClose={handleNewProviderClose} addProvider={handleAddProvider}/>
			<DeleteDialog open={deleteProviderOpen} handleClose={handleDeleteProviderClose} deleteProvider={handleDeleteProvider} />
		</div>
	);
}

export default Dashboard;
