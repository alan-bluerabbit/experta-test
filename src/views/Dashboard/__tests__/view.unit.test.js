import React from 'react';
import renderer from 'react-test-renderer';
import Dashboard from '../view';
import { useStore } from '../../../helpers/store';

jest.mock("../../../helpers/store");

useStore.mockReturnValue({
    state: {
        providers: [
            {
                id: 123456,
                name: 'testName',
                mail: 'test@mail.com',
                cuit: 20-12345678-7,
                address: 'test address 123',
                phone: '+5491151234567'
            }
        ]
    }
})


test('<Dashboard />', () => {
    const tree = renderer
    .create(<Dashboard />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});