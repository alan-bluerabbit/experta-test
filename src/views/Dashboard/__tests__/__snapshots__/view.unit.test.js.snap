// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> 1`] = `
<div
  className="makeStyles-root-1"
>
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionAbsolute MuiAppBar-colorPrimary makeStyles-appBar-3 MuiPaper-elevation4"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-2 MuiToolbar-gutters"
    >
      <h1
        className="MuiTypography-root makeStyles-title-4 MuiTypography-h6 MuiTypography-colorInherit MuiTypography-noWrap"
      >
        Experta Seguros
      </h1>
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
      </button>
    </div>
  </header>
  <main
    className="makeStyles-content-6"
  >
    <div
      className="makeStyles-appBarSpacer-5"
    />
    <div
      className="MuiContainer-root makeStyles-container-7 MuiContainer-maxWidthLg"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            className="MuiPaper-root makeStyles-paper-8 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <h2
              className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
            >
              Proveedores
            </h2>
            <table
              className="MuiTable-root"
              role={null}
            >
              <thead
                className="MuiTableHead-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root MuiTableRow-head"
                  role={null}
                >
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                    scope="col"
                  >
                    Nombre
                  </th>
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                    scope="col"
                  >
                    CUIT
                  </th>
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
                    scope="col"
                  >
                    Fecha de creación
                  </th>
                  <th
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
                    scope="col"
                  />
                </tr>
              </thead>
              <tbody
                className="MuiTableBody-root"
                role={null}
              >
                <tr
                  className="MuiTableRow-root"
                  role={null}
                >
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                  >
                    testName
                  </td>
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                  >
                    -12345665
                  </td>
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall"
                  >
                    Invalid Date
                  </td>
                  <td
                    aria-sort={null}
                    className="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeSmall"
                  >
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-text"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root"
                          focusable="false"
                          style={
                            Object {
                              "color": "grey",
                            }
                          }
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                          />
                        </svg>
                      </span>
                    </button>
                    <button
                      className="MuiButtonBase-root MuiButton-root MuiButton-text"
                      disabled={false}
                      onBlur={[Function]}
                      onClick={[Function]}
                      onDragLeave={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      onKeyUp={[Function]}
                      onMouseDown={[Function]}
                      onMouseLeave={[Function]}
                      onMouseUp={[Function]}
                      onTouchEnd={[Function]}
                      onTouchMove={[Function]}
                      onTouchStart={[Function]}
                      tabIndex={0}
                      type="button"
                    >
                      <span
                        className="MuiButton-label"
                      >
                        <svg
                          aria-hidden={true}
                          className="MuiSvgIcon-root"
                          focusable="false"
                          style={
                            Object {
                              "color": "grey",
                            }
                          }
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                          />
                        </svg>
                      </span>
                    </button>
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              className="makeStyles-seeMore-9"
            >
              <a
                className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="#"
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
              >
                Ver más proveedores
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiBox-root MuiBox-root-10"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
        >
          Copyright © 
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
            href="https://www.experta.com.ar"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            Experta Seguros
          </a>
           
          2020
          .
        </p>
      </div>
    </div>
  </main>
</div>
`;
